File: src/main/folder-prefs.ts
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs'
import { join, basename, dirname } from 'path'
import { homedir } from 'os'
import type { ClaudeSettings, ClaudeMdFile, SettingsFile, FolderPreferences, McpServerInfo, McpServersResult } from '../shared/types'

const home = homedir()
const claudeDir = join(home, '.claude')

function readJson(path: string): ClaudeSettings {
  try {
    return JSON.parse(readFileSync(path, 'utf-8'))
  } catch {
    return {}
  }
}

function readText(path: string): string {
  try {
    return readFileSync(path, 'utf-8')
  } catch {
    return ''
  }
}

export function loadFolderPreferences(projectPath: string): FolderPreferences {
  const managedPath = join(claudeDir, 'managed-settings.json')
  const globalPath = join(claudeDir, 'settings.json')
  const globalLocalPath = join(claudeDir, 'settings.local.json')
  const projectSettingsPath = join(projectPath, '.claude', 'settings.json')
  const projectLocalPath = join(projectPath, '.claude', 'settings.local.json')

  const settingsFiles: SettingsFile[] = [
    {
      path: managedPath,
      label: 'Managed (Organization)',
      scope: 'managed',
      readOnly: true,
      exists: existsSync(managedPath),
      settings: readJson(managedPath),
    },
    {
      path: globalPath,
      label: 'Global User',
      scope: 'global',
      readOnly: false,
      exists: existsSync(globalPath),
      settings: readJson(globalPath),
    },
    {
      path: globalLocalPath,
      label: 'Global Local',
      scope: 'global-local',
      readOnly: false,
      exists: existsSync(globalLocalPath),
      settings: readJson(globalLocalPath),
    },
    {
      path: projectSettingsPath,
      label: 'Project (shared)',
      scope: 'project',
      readOnly: false,
      exists: existsSync(projectSettingsPath),
      settings: readJson(projectSettingsPath),
    },
    {
      path: projectLocalPath,
      label: 'Project Local',
      scope: 'project-local',
      readOnly: false,
      exists: existsSync(projectLocalPath),
      settings: readJson(projectLocalPath),
    },
  ]

  const globalMdPath = join(claudeDir, 'CLAUDE.md')
  const projectMdPath = join(projectPath, 'CLAUDE.md')
  const projectNestedMdPath = join(projectPath, '.claude', 'CLAUDE.md')

  const mdFiles: ClaudeMdFile[] = [
    {
      path: globalMdPath,
      label: 'Global (~/.claude/CLAUDE.md)',
      scope: 'global',
      exists: existsSync(globalMdPath),
      content: readText(globalMdPath),
    },
    {
      path: projectNestedMdPath,
      label: 'Project (.claude/CLAUDE.md)',
      scope: 'project-nested',
      exists: existsSync(projectNestedMdPath),
      content: readText(projectNestedMdPath),
    },
    {
      path: projectMdPath,
      label: 'Project Root (CLAUDE.md)',
      scope: 'project',
      exists: existsSync(projectMdPath),
      content: readText(projectMdPath),
    },
  ]

  return {
    projectPath,
    projectName: basename(projectPath),
    settingsFiles,
    mdFiles,
  }
}

export function saveSettingsFile(filePath: string, updates: ClaudeSettings): void {
  // Re-read and merge to avoid clobbering concurrent writes (e.g. Operator hooks)
  const existing = existsSync(filePath) ? readJson(filePath) : {}
  const merged = { ...existing, ...updates }

  const dir = dirname(filePath)
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  writeFileSync(filePath, JSON.stringify(merged, null, 2) + '\n', 'utf-8')
}

export function saveMdFile(filePath: string, content: string): void {
  const dir = dirname(filePath)
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  writeFileSync(filePath, content, 'utf-8')
}

export function createFile(filePath: string, type: 'settings' | 'md'): void {
  const dir = dirname(filePath)
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  const content = type === 'settings' ? '{}\n' : ''
  writeFileSync(filePath, content, 'utf-8')
}

export function getMcpServers(projectPath: string): McpServersResult {
  const servers: McpServerInfo[] = []

  // Read from ~/.claude.json (user-level MCP servers)
  const userClaudeJson = join(home, '.claude.json')
  try {
    const data = JSON.parse(readFileSync(userClaudeJson, 'utf-8'))
    if (data.mcpServers && typeof data.mcpServers === 'object') {
      for (const [name, config] of Object.entries(data.mcpServers)) {
        const c = config as Record<string, unknown>
        servers.push({
          name,
          type: (c.type as 'stdio' | 'http') || 'stdio',
          source: '~/.claude.json',
        })
      }
    }
    // Cloud MCPs (claude.ai integrations)
    if (Array.isArray(data.claudeAiMcpEverConnected)) {
      for (const name of data.claudeAiMcpEverConnected) {
        servers.push({ name, type: 'cloud', source: 'cloud' })
      }
    }
  } catch {
    // No user config or parse error
  }

  // Read from project .mcp.json
  const projectMcpJson = join(projectPath, '.mcp.json')
  try {
    const data = JSON.parse(readFileSync(projectMcpJson, 'utf-8'))
    if (data.mcpServers && typeof data.mcpServers === 'object') {
      for (const [name, config] of Object.entries(data.mcpServers)) {
        const c = config as Record<string, unknown>
        servers.push({
          name,
          type: (c.type as 'stdio' | 'http') || 'stdio',
          source: '.mcp.json',
        })
      }
    }
  } catch {
    // No project MCP config
  }

  return { servers }
}

File: src-tauri/src/folderprefs.rs
// Folder preferences (port of folder-prefs.ts): read/write Claude Code settings
// JSON + CLAUDE.md files at global/project scope, and discover MCP servers.

use std::path::{Path, PathBuf};

use serde::Serialize;
use serde_json::{json, Value};

fn claude_dir() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_default();
    PathBuf::from(home).join(".claude")
}

fn read_json(path: &Path) -> Value {
    std::fs::read_to_string(path).ok().and_then(|s| serde_json::from_str(&s).ok()).unwrap_or_else(|| json!({}))
}

fn read_text(path: &Path) -> String {
    std::fs::read_to_string(path).unwrap_or_default()
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SettingsFile {
    pub path: String,
    pub label: String,
    pub scope: String,
    pub read_only: bool,
    pub exists: bool,
    pub settings: Value,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaudeMdFile {
    pub path: String,
    pub label: String,
    pub scope: String,
    pub exists: bool,
    pub content: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderPreferences {
    pub project_path: String,
    pub project_name: String,
    pub settings_files: Vec<SettingsFile>,
    pub md_files: Vec<ClaudeMdFile>,
}

fn sf(path: PathBuf, label: &str, scope: &str, read_only: bool) -> SettingsFile {
    let exists = path.exists();
    SettingsFile { settings: read_json(&path), path: path.to_string_lossy().to_string(), label: label.into(), scope: scope.into(), read_only, exists }
}

fn global_settings_files() -> Vec<SettingsFile> {
    let d = claude_dir();
    vec![
        sf(d.join("managed-settings.json"), "Managed (Organization)", "managed", true),
        sf(d.join("settings.json"), "Global User", "global", false),
        sf(d.join("settings.local.json"), "Global Local", "global-local", false),
    ]
}

fn md(path: PathBuf, label: &str, scope: &str) -> ClaudeMdFile {
    let exists = path.exists();
    ClaudeMdFile { content: read_text(&path), path: path.to_string_lossy().to_string(), label: label.into(), scope: scope.into(), exists }
}

fn global_md_files() -> Vec<ClaudeMdFile> {
    vec![md(claude_dir().join("CLAUDE.md"), "Global (~/.claude/CLAUDE.md)", "global")]
}

pub fn load_global_preferences() -> FolderPreferences {
    FolderPreferences {
        project_path: claude_dir().to_string_lossy().to_string(),
        project_name: "Global Claude".into(),
        settings_files: global_settings_files(),
        md_files: global_md_files(),
    }
}

pub fn load_folder_preferences(project_path: &str) -> FolderPreferences {
    let p = Path::new(project_path);
    let mut settings_files = global_settings_files();
    settings_files.push(sf(p.join(".claude").join("settings.json"), "Project (shared)", "project", false));
    settings_files.push(sf(p.join(".claude").join("settings.local.json"), "Project Local", "project-local", false));

    let mut md_files = global_md_files();
    md_files.push(md(p.join(".claude").join("CLAUDE.md"), "Project (.claude/CLAUDE.md)", "project-nested"));
    md_files.push(md(p.join("CLAUDE.md"), "Project Root (CLAUDE.md)", "project"));

    FolderPreferences {
        project_path: project_path.to_string(),
        project_name: p.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_else(|| project_path.to_string()),
        settings_files,
        md_files,
    }
}

pub fn save_settings_file(path: &str, updates: Value) {
    let p = Path::new(path);
    let mut existing = if p.exists() { read_json(p) } else { json!({}) };
    if let (Some(base), Some(up)) = (existing.as_object_mut(), updates.as_object()) {
        for (k, v) in up {
            base.insert(k.clone(), v.clone());
        }
    } else {
        existing = updates;
    }
    if let Some(dir) = p.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    if let Ok(s) = serde_json::to_string_pretty(&existing) {
        let _ = std::fs::write(p, format!("{s}\n"));
    }
}

pub fn save_md_file(path: &str, content: &str) -> Result<(), String> {
    let p = Path::new(path);
    if let Some(dir) = p.parent() {
        std::fs::create_dir_all(dir).map_err(|e| format!("create {}: {e}", dir.display()))?;
    }
    std::fs::write(p, content).map_err(|e| format!("write {}: {e}", p.display()))
}

pub fn create_file(path: &str, kind: &str) {
    let p = Path::new(path);
    if let Some(dir) = p.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    let content = if kind == "settings" { "{}\n" } else { "" };
    let _ = std::fs::write(p, content);
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServerInfo {
    pub name: String,
    #[serde(rename = "type")]
    pub kind: String,
    pub source: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServersResult {
    pub servers: Vec<McpServerInfo>,
}

fn collect_mcp(data: &Value, source: &str, out: &mut Vec<McpServerInfo>) {
    if let Some(map) = data.get("mcpServers").and_then(|v| v.as_object()) {
        for (name, cfg) in map {
            let kind = cfg.get("type").and_then(|v| v.as_str()).unwrap_or("stdio").to_string();
            out.push(McpServerInfo { name: name.clone(), kind, source: source.into() });
        }
    }
}

pub fn get_mcp_servers(project_path: &str) -> McpServersResult {
    let home = std::env::var("HOME").unwrap_or_default();
    let mut servers = vec![];

    let user_json = read_json(&Path::new(&home).join(".claude.json"));
    collect_mcp(&user_json, "~/.claude.json", &mut servers);
    if let Some(cloud) = user_json.get("claudeAiMcpEverConnected").and_then(|v| v.as_array()) {
        for name in cloud.iter().filter_map(|v| v.as_str()) {
            servers.push(McpServerInfo { name: name.to_string(), kind: "cloud".into(), source: "cloud".into() });
        }
    }
    collect_mcp(&read_json(&Path::new(project_path).join(".mcp.json")), ".mcp.json", &mut servers);

    McpServersResult { servers }
}
